docs/versioning.md

```diff
@@ -0,0 +1,488 @@
+
+To work with all existing versions of an object, S3 provides an extended
+API which uses the same operations of the standard API, but with versioning
+information specified. For example, in order to get the version `v1` of `foo`,
+users have to specify that version in the query, such as
+`GET /foo?versionId=v1`.
+
+A bucket can be non-versioning or versioning. Any object placed in a bucket
+before versioning is enabled is considered to be the "null" version. When
+versioning is enabled, the "null" version will be kept and continue to be
+identified as the "null" version. However, if versioning is suspended, the
+"null" version may be overwritten with a new "null" version. All versions with
+a version ID other than "null" are retained when versioning is suspended, even
+when there are new puts on the object.
+
+## Implementation of Bucket Versioning in S3
+
+### Roles of Metadata Engine and API Component
+
+Each version of an object is stored as a separate key in metadata. The API
+component of S3 manages the put and delete of version data, sends instructions
+to create, retrieve, overwrite or delete a version's metadata to the metadata
+backend, and tracks extra information about a version, such as whether it is a
+delete marker or a null version.
+
+The metadata backend is responsible for handling the create, overwrite and
+delete operations of version metadata upon receiving instructions. We use
+version identifiers as the suffix for the keys of the object versions, and a
+special version to represent the latest version.
+
+This is an example of what the keys may look like in metadata, for an object
+`foo/bar` with three versions. The latest version is represented below in the
+key `foo/bar`. This special version is called the master version and is
+described further in one of the following sections, "Master Version".
+
+| key |
+|-----|
+| foo/bar |
+| foo/bar\098506163554375999999PARIS  0.a430a1f85c6ec |
+| foo/bar\098506163554373999999PARIS  0.41b510cd0fdf8 |
+| foo/bar\098506163554373999998PARIS  0.f9b82c166f695 |
+
+The following is a summary of the design and implementation details of bucket
+versioning, described first for the metadata engine and then for the API.
+
+## Implementation of Bucket Versioning in Metadata
+
+### Version Id Format
+
+The version id is generated by the metadata backend, and encoded in a
+hexadecimal string format by S3. The following information is encoded before
+sending a response to a request. S3 also decodes the hexadecimal string
+received from a request before sending to metadata to retrieve a particular
+version.
+
+`object_name separator version_id`
+where:
+
+- `object_name`: is the key of the object in metadata
+- `separator`: we use the `null` character (`0x00` or `\0`) as the separator
+  between the `object_name` and the `version_id` of a key
+- `version_id`: is the version identifier; this encodes the ordering
+  information in the texture format as metadata orders keys alphabetically
```

Comment:
I mentioned before that it was in the original document. I assumed it meant the format used for the version ID described below. Since it's not clear, I removed the term.